File: packages/newsletter-generator/examples/basic-example.ts
import { NewsletterGenerator, NewsletterRenderer } from '../src';
import type { GenerateParams, MediaAnalysis } from '../src/types';
import * as dotenv from 'dotenv';

// .envファイルを読み込み
dotenv.config();

// サンプルデータを作成
const createSampleData = (): GenerateParams => {
  const now = new Date();
  const mediaAnalyses: MediaAnalysis[] = [
    {
      mediaId: 'photo-001',
      filePath: 'gs://bucket/photos/photo-001.jpg',
      type: 'photo',
      capturedAt: new Date('2024-01-15'),
      expressions: [
        { type: 'smile', confidence: 0.95 }
      ],
      actions: [
        { type: 'sitting', confidence: 0.88 }
      ],
      objects: [
        { name: 'toy_blocks', category: 'toy', confidence: 0.92 }
      ]
    },
    {
      mediaId: 'photo-002',
      filePath: 'gs://bucket/photos/photo-002.jpg',
      type: 'photo',
      capturedAt: new Date('2024-01-20'),
      expressions: [
        { type: 'laugh', confidence: 0.90 }
      ],
      actions: [
        { type: 'standing', confidence: 0.85 }
      ],
      objects: [
        { name: 'ball', category: 'toy', confidence: 0.88 }
      ]
    },
    {
      mediaId: 'video-001',
      filePath: 'gs://bucket/videos/video-001.mp4',
      type: 'video',
      capturedAt: new Date('2024-01-25'),
      expressions: [
        { type: 'smile', confidence: 0.87, timestamp: 2.5 }
      ],
      actions: [
        { type: 'walking', confidence: 0.92, timestamp: 5.0 }
      ],
      videoSummary: '公園でボール遊びを楽しんでいる様子',
      importantScenes: [
        {
          startTime: 3.0,
          endTime: 7.0,
          description: '初めてボールをキックできた瞬間',
          significance: 'high'
        }
      ]
    }
  ];

  return {
    childProfile: {
      id: 'child-123',
      name: '太郎',
      birthDate: new Date('2022-06-01'),
      gender: 'male',
      currentAge: {
        years: 1,
        months: 7
      }
    },
    period: {
      start: new Date('2024-01-01'),
      end: new Date('2024-01-31')
    },
    mediaAnalyses,
    timeline: {
      childId: 'child-123',
      milestones: [
        {
          id: 'milestone-001',
          type: 'first_walk',
          date: new Date('2023-12-15'),
          description: '初めて一人で歩いた',
          mediaIds: []
        }
      ],
      preferences: [
        {
          category: 'toy',
          items: ['ブロック', 'ボール'],
          lastObserved: new Date('2024-01-20')
        }
      ],
      lastUpdated: new Date()
    }
  };
};

async function main() {
  console.log('連絡帳生成のサンプルを実行します...\n');

  // .envファイルまたは環境変数からプロジェクトIDを取得
  const projectId = process.env.GCP_PROJECT_ID || 'your-project-id';
  const location = process.env.GCP_LOCATION || 'asia-northeast1';
  
  if (projectId === 'your-project-id') {
    console.warn('⚠️  GCP_PROJECT_IDが設定されていません。');
    console.log('\n設定方法:');
    console.log('1. .envファイルを作成（推奨）:');
    console.log('   cp .env.example .env');
    console.log('   # .envファイルを編集してGCP_PROJECT_IDを設定');
    console.log('\n2. または環境変数で設定:');
    console.log('   export GCP_PROJECT_ID=your-actual-project-id\n');
  }

  try {
    // 生成器を初期化
    const generator = new NewsletterGenerator(projectId, location);
    const renderer = new NewsletterRenderer();

    // サンプルデータを作成
    const params = createSampleData();

    console.log('📝 連絡帳を生成中...');
    const newsletter = await generator.generate(params);
    
    console.log('\n✅ 連絡帳が生成されました:');
    console.log(`- ID: ${newsletter.id}`);
    console.log(`- タイトル: ${newsletter.title}`);
    console.log(`- セクション数: ${newsletter.sections.length}`);
    console.log(`- 期間: ${newsletter.period.start.toLocaleDateString()} ～ ${newsletter.period.end.toLocaleDateString()}`);

    // HTMLとして出力
    const outputPath = './examples/output';
    console.log('\n📄 ファイルを出力中...');
    const outputs = await renderer.renderAll(newsletter, outputPath, ['html']);
    
    console.log('\n✅ 出力完了:');
    Object.entries(outputs).forEach(([format, path]) => {
      console.log(`- ${format.toUpperCase()}: ${path}`);
    });

    // 再生成のサンプル
    console.log('\n🔄 プロンプトで再生成中...');
    
    // 初回生成から少し間隔を空ける
    console.log('⏳ レート制限回避のため5秒待機中...');
    await new Promise(resolve => setTimeout(resolve, 5000));
    
    const regenerated = await generator.regenerate({
      newsletter,
      prompt: 'もっと具体的な成長の様子を詳しく記載してください'
    });
    
    console.log('✅ 再生成完了 (バージョン: ' + regenerated.version + ')');

  } catch (error) {
    console.error('\n❌ エラーが発生しました:', error);
    if (error instanceof Error && error.message.includes('Application Default Credentials')) {
      console.log('\nGCP認証の設定が必要です。以下のいずれかの方法で設定してください:');
      console.log('1. gcloud auth application-default login');
      console.log('2. サービスアカウントキーを使用');
      console.log('   export GOOGLE_APPLICATION_CREDENTIALS=path/to/key.json');
    }
  }
}

// 実行
main().catch(console.error);

File: packages/newsletter-generator/examples/basic_example.py
"""Basic example of newsletter generation."""

import asyncio
import os
from datetime import datetime
from pathlib import Path

from dotenv import load_dotenv

from newsletter_generator import (
    ChildProfile,
    GenerateParams,
    MockRecordReader,
    NewsletterExporter,
    NewsletterGenerator,
    create_sample_childcare_records,
)


def create_sample_data() -> GenerateParams:
    """サンプルデータを作成する."""
    # モック育児記録リーダーを作成
    mock_reader = MockRecordReader()
    
    # サンプル育児記録を追加
    sample_records = create_sample_childcare_records()
    mock_reader.add_records(sample_records)
    
    return GenerateParams(
        child_profile=ChildProfile(
            id="child-123",
            name="太郎",
            birth_date=datetime(2022, 6, 1),
            gender="male",
        ),
        period={
            "start": datetime(2024, 1, 1),
            "end": datetime(2024, 1, 31),
        },
        record_reader=mock_reader,
        timeline={
            "child_id": "child-123",
            "milestones": [
                {
                    "id": "milestone-001",
                    "type": "first_walk",
                    "date": "2023-12-15",
                    "description": "初めて一人で歩いた",
                    "media_ids": [],
                }
            ],
            "preferences": [
                {
                    "category": "toy",
                    "items": ["ブロック", "ボール"],
                    "last_observed": datetime(2024, 1, 20),
                }
            ],
            "last_updated": datetime.now(),
        },
    )


async def main() -> None:
    """メイン処理."""
    print("連絡帳生成のサンプルを実行します...\n")
    
    # 環境変数を読み込み
    load_dotenv()
    
    # API キーの確認
    api_key = os.getenv("GOOGLE_API_KEY")
    if not api_key:
        print("⚠️  GOOGLE_API_KEY が設定されていません。")
        print("\n設定方法:")
        print("1. .env ファイルを作成（推奨）:")
        print("   cp .env.example .env")
        print("   # .env ファイルを編集して GOOGLE_API_KEY を設定")
        print("\n2. または環境変数で設定:")
        print("   export GOOGLE_API_KEY=your-actual-api-key\n")
        return
    
    try:
        # 生成器を初期化
        generator = NewsletterGenerator()
        exporter = NewsletterExporter()
        
        # サンプルデータを作成
        params = create_sample_data()
        
        print("📝 連絡帳を生成中...")
        print("- 育児記録から情報を読み込み中...")
        newsletter = await generator.generate(params)
        
        print("\n✅ 連絡帳が生成されました:")
        print(f"- ID: {newsletter.id}")
        print(f"- タイトル: {newsletter.title}")
        print(f"- セクション数: {len(newsletter.sections)}")
        print(
            f"- 期間: {newsletter.period['start'].strftime('%Y/%m/%d')} ～ "
            f"{newsletter.period['end'].strftime('%Y/%m/%d')}"
        )
        print(f"- 使用した育児記録数: {newsletter.metadata.get('record_count', 0)}件")
        
        # JSONとして出力
        output_path = Path("./examples/output")
        output_path.mkdir(parents=True, exist_ok=True)
        
        print("\n📄 JSONファイルを出力中...")
        json_path = output_path / f"newsletter_{newsletter.id}.json"
        exporter.save_json(newsletter, json_path)
        
        print(f"\n✅ 出力完了:")
        print(f"- JSON: {json_path}")
        
        # コンソールにも構造を表示
        print("\n📋 生成されたコンテンツ構造:")
        newsletter_dict = exporter.to_dict(newsletter)
        for section in newsletter_dict["sections"]:
            print(f"\n【{section['title']}】(type: {section['type']})")
            content = section["content"]
            if content.get("text"):
                print(f"  テキスト: {content['text'][:100]}...")
            if content.get("caption"):
                print(f"  キャプション: {content['caption']}")
            if content.get("photo_url"):
                print(f"  写真URL: {content['photo_url']}")
        
        # 再生成のサンプル（レート制限を考慮してコメントアウト）
        # print("\n🔄 プロンプトで再生成中...")
        # 
        # # 初回生成から少し間隔を空ける
        # print("⏳ レート制限回避のため5秒待機中...")
        # await asyncio.sleep(5)
        # 
        # regenerated = await generator.regenerate(
        #     RegenerateParams(
        #         newsletter=newsletter,
        #         prompt="育児記録の具体的な観察内容をより詳しく記載してください",
        #     )
        # )
        # 
        # print(f"✅ 再生成完了 (バージョン: {regenerated.version})")
        
    except Exception as e:
        print(f"\n❌ エラーが発生しました: {e}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    asyncio.run(main())